### PR TITLE
Issue 331/mention support on @ keypress

### LINKED ISSUE #331: Support @-mentions while editing
It would be nice if we could support @-mentions while editing, to add an explicit mention to a site user.

### PR DESCRIPTION
**Pull request by SergioEstevao**
Fixes #331

Related PRs:

 - Gutenberg PR: https://github.com/WordPress/gutenberg/pull/22119
 - WP-iOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/14088
 - WP-Android: https://github.com/wordpress-mobile/WordPress-Android/pull/12115

This PR changes the AztecView component to be able to intercept any keycode that is defined. It then configures uses this functionality to intercept the `@` keypress and start the mention UI.

To test:
 - Start the demo app
 - Add a Paragraph block
 - Tap the @ on the keyboard
 - Check if a mention to `matt` is added
 - Check if the mention is only inserted if the @ is inserted with a space before or at the start of the paragraph.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: react-native-aztec/src/AztecView.js [hunk 1]
```diff
@@ -165,14 +181,15 @@ class AztecView extends React.Component {
           onContentSizeChange = { this._onContentSizeChange }
           onHTMLContentWithCursor = { this._onHTMLContentWithCursor }
           onSelectionChange = { this._onSelectionChange }
-          onEnter = { this.props.onEnter && this._onEnter }
+          onEnter = { this.props.onKeyDown && this._onEnter }
+          onBackspace = { this.props.onKeyDown && this._onBackspace }
```
Comment (SergioEstevao) [inline react-native-aztec/src/AztecView.js:185]: After the Android Aztec lib code is updated the `onEnter` and `onBackspace` props should be remove, and all "keypress" events should go troigh `onKeyDown`

**Comment by SergioEstevao** [comment]
Tested this on Android and is working great!

@iamthomasbishop do you mind review the Android version to see if the design of the mentions view UI/UX is good?

File: react-native-aztec/src/AztecView.js [hunk 2]
```diff
@@ -61,28 +63,42 @@ class AztecView extends React.Component {
...
 
   _onBackspace = (event) => {
-    if (!this.props.onBackspace) {
+    if (!this.props.onKeyDown) {
+      return;
+    }
+
+    const { onKeyDown } = this.props;
+
+    let newEvent = { ...event, keyCode: 8, preventDefault: () => {} }
```
Comment (SergioEstevao) [inline react-native-aztec/src/AztecView.js:83]: Maybe define a constant that translates the "magic number" 8 ?

**Comment by iamthomasbishop** [comment]
@mchowning @SergioEstevao I had another chance to test on Android in more detail. Here are some notes:

* I'm not seeing the `@` icon-button in the block toolbar — not sure if this was intentionally removed, but I would like to keep it
* There is currently a short fadeOut animation on the text input row when tapping on a result from the suggestions list. Can we hide the text input row immediately?
* There's a subtle "flash" on the keyboard when you tap a suggestion ([video](https://share.icloud.com/photos/0saxnY96Fmp1tInRurfBn1_FA))
* Can we add a 1px border-top to the text input row, using the same border color as the block toolbar? Think we had that at some point but it must’ve been removed while we iterated style.

I’d also I’ll Ike to cross-check on iOS to see if any of these issues exist on iOS (particularly the first one). @SergioEstevao are you able to make a new test build?

**Comment by SergioEstevao** [comment]
@thomasbishop The @ button was missing in the toolbar because it was hidden
because of the DEV flag settings.

I will remove the flag and generate another build for your to check.

**Comment by iamthomasbishop** [comment]
@SergioEstevao Tested iOS and it is looking solid — not seeing the aforementioned issues there. 👍

**Comment by mchowning** [comment]
:wave: @iamthomasbishop . Thanks for taking a look. :bow: I've addressed your comments and there's [a new build you can test](https://github.com/wordpress-mobile/WordPress-Android/pull/12115#issuecomment-639497830). Care to take another look and let me know how things are looking now?

> I'm not seeing the @ icon-button in the block toolbar — not sure if this was intentionally removed, but I would like to keep it

As [Sergio noted](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2218#issuecomment-643484222), that was because we had kept the button behind a dev flag until now. That has been removed in the latest code though.

> There is currently a short fadeOut animation on the text input row when tapping on a result from the suggestions list. Can we hide the text input row immediately?
>...
> Can we add a 1px border-top to the text input row, using the same border color as the block toolbar? Think we had that at some point but it must’ve been removed while we iterated style.

Removed the animation and added the top border back.

> There's a subtle "flash" on the keyboard when you tap a suggestion (video)

Also took care of this. There is still a small keyboard flash when opening the mentions UI, but it's quite small (at least in my testing) and I think resolving that flash will probably be difficult. If this is something we have to fix before we can release this though, just let me know.

**Comment by iamthomasbishop** [comment]
@mchowning Most everything is looking solid, just a couple of things.

- The border on the input row doesn't match the block toolbar in light mode ([example](https://cloudup.com/cCXjm8yFPrZ) of the color it should be). There shouldn't be a border in dark mode (or it can be transparent)
- In dark mode, the background color should match the block toolbar background ([example](https://cloudup.com/cPJdZ7cyehV))
- I noticed that when you start typing and the suggestions list narrows down from many to 1, the "no matching users available" label is still visible behind the UI. Try typing "matt", waiting then typing "c" — when the results narrow to 1, you'll see the label still visible behind it

I'm not seeing that "selected" flash anymore, nice fix! I'm also not seeing any strong flashing while opening the mentions UI, at least during this quick test.

**Comment by mchowning** [comment]
> @mchowning Most everything is looking solid, just a couple of things...

Thanks @iamthomasbishop. I believe I've addressed those issues, and [there's a new build available to test](https://github.com/wordpress-mobile/WordPress-Android/pull/12115#issuecomment-639497830). Let me know what you think!

**Comment by iamthomasbishop** [comment]
@mchowning had a chance to take the latest build for a spin and it's looking solid. I think we are in good shape to ship it! 😁

**Comment by SergioEstevao** [comment]
Closing this one in favor of https://github.com/wordpress-mobile/gutenberg-mobile/pull/2424 that takes into account the new monorepo structure of the repo.
All the changes that were made in this PR were move to this PR in GB: https://github.com/WordPress/gutenberg/pull/22119

**Pull request closed without merging**